typescript/core/src/parse.ts

```diff
@@ -1,374 +1,415 @@
+  const start = reader.offset;
+  const opcode = reader.uint8();
+  const recordLength = reader.uint64();
 
-  const opcode = headerReader.uint8();
-
-  const recordLength = headerReader.uint64();
   if (recordLength > Number.MAX_SAFE_INTEGER) {
     throw new Error(`Record content length ${recordLength} is too large`);
   }
+
   const recordLengthNum = Number(recordLength);
-  const recordEndOffset = headerReader.offset + recordLengthNum;
-  if (recordEndOffset > view.byteLength) {
-    return { usedBytes: 0 };
+
+  if (reader.bytesRemaining() < recordLengthNum) {
+    reader.offset = start; // Rewind to the start of the record
+    return undefined;
   }
 
-  if (!isKnownOpcode(opcode)) {
-    const data = new Uint8Array(
-      view.buffer,
-      view.byteOffset + headerReader.offset,
-      recordLengthNum,
-    );
-    const record: TypedMcapRecord = {
-      type: "Unknown",
-      opcode,
-      data,
-    };
-    return { record, usedBytes: recordEndOffset - startOffset };
+  let result: TypedMcapRecord;
+  switch (opcode as Opcode) {
+    case Opcode.HEADER:
+      result = parseHeader(reader, recordLengthNum);
+      break;
+    case Opcode.FOOTER:
+      result = parseFooter(reader, recordLengthNum);
+      break;
+    case Opcode.SCHEMA:
+      result = parseSchema(reader, recordLengthNum);
+      break;
+    case Opcode.CHANNEL:
+      result = parseChannel(reader, recordLengthNum);
+      break;
+    case Opcode.MESSAGE:
+      result = parseMessage(reader, recordLengthNum);
+      break;
+    case Opcode.CHUNK:
+      result = parseChunk(reader, recordLengthNum);
+      break;
+    case Opcode.MESSAGE_INDEX:
+      result = parseMessageIndex(reader, recordLengthNum);
+      break;
+    case Opcode.CHUNK_INDEX:
+      result = parseChunkIndex(reader, recordLengthNum);
+      break;
+    case Opcode.ATTACHMENT:
+      result = parseAttachment(reader, recordLengthNum, validateCrcs);
+      break;
+    case Opcode.ATTACHMENT_INDEX:
+      result = parseAttachmentIndex(reader, recordLengthNum);
+      break;
+    case Opcode.STATISTICS:
+      result = parseStatistics(reader, recordLengthNum);
+      break;
+    case Opcode.METADATA:
+      result = parseMetadata(reader, recordLengthNum);
+      break;
+    case Opcode.METADATA_INDEX:
+      result = parseMetadataIndex(reader, recordLengthNum);
+      break;
+    case Opcode.SUMMARY_OFFSET:
+      result = parseSummaryOffset(reader, recordLengthNum);
+      break;
+    case Opcode.DATA_END:
+      result = parseDataEnd(reader, recordLengthNum);
+      break;
+    default:
+      result = parseUnknown(reader, recordLengthNum, opcode);
+      break;
   }
 
-  const recordView = new DataView(
-    view.buffer,
-    view.byteOffset + headerReader.offset,
-    recordLengthNum,
+  // NOTE: a bit redundant, but ensures we've advanced by the full record length
+  // TODO: simplify this when we explore monomorphic paths
+  reader.offset = start + RECORD_HEADER_SIZE + recordLengthNum;
```

Comment:
should this be more like an assert()? or is there some legitimate case today where this is not already at the right offset?